Mathlib/Analysis/Meromorphic/Divisor/Basic.lean

```diff
@@ -0,0 +1,424 @@
+
+/-- The condition `supportDiscreteWithinU` in a divisor is equivalent to saying
+that the support is locally finite near every point of `U`. -/
+theorem supportDiscreteWithin_iff_locallyFiniteWithin {f : 𝕜 → ℤ} (h : f.support ⊆ U) :
+    f =ᶠ[codiscreteWithin U] 0 ↔ ∀ z ∈ U, ∃ t ∈ 𝓝 z, Set.Finite (t ∩ f.support) := by
+  have : f.support = (U \ {x | f x = (0 : 𝕜 → ℤ) x}) := by
+    ext x
+    simp only [Function.mem_support, ne_eq, Pi.zero_apply, mem_diff, mem_setOf_eq, iff_and_self]
+    exact (h ·)
+  rw [EventuallyEq, Filter.Eventually, codiscreteWithin_iff_locallyFiniteComplementWithin, this]
+
+namespace Divisor
+
+/-- A divisor can be coerced into a function 𝕜 → ℤ -/
+instance (U : Set 𝕜) : CoeFun (Divisor U) (fun _ ↦ 𝕜 → ℤ) where
+  coe := Divisor.toFun
+
+attribute [coe] Divisor.toFun
+
+/-- Divisors are `FunLike`: the coercion from divisors to functions is injective. -/
+instance : FunLike (Divisor U) 𝕜 ℤ where
+  coe := fun D ↦ D.toFun
+  coe_injective' := fun ⟨_, _, _⟩ ⟨_, _, _⟩ ↦ by simp
+
+/-- Helper lemma for the `ext` tactic: two divisors are equal if their
+associated functions agree. -/
+@[ext]
+theorem ext {D₁ D₂ : Divisor U} (h : ∀ a, D₁.toFun a = D₂.toFun a) : D₁ = D₂ := DFunLike.ext _ _ h
+
+/-!
+## Elementary properties of the support
+-/
+
+/-- The support of a divisor is discrete. -/
+theorem discreteSupport (D : Divisor U) :
+    DiscreteTopology D.toFun.support := by
+  have : Function.support D = {x | D x = 0}ᶜ ∩ U := by
+    ext x
+    constructor
+    · exact fun hx ↦ ⟨by tauto, D.supportWithinDomain hx⟩
+    · intro hx
+      rw [mem_inter_iff, mem_compl_iff, mem_setOf_eq] at hx
+      tauto
+  convert discreteTopology_of_codiscreteWithin (D.supportDiscreteWithinDomain)
+
+/-- If `U` is closed, the the support of a divisor on `U` is also closed. -/
+theorem closedSupport (D : Divisor U) (hU : IsClosed U) :
+    IsClosed D.toFun.support := by
+  convert closed_compl_of_codiscreteWithin D.supportDiscreteWithinDomain hU
+  ext x
+  constructor
+  · intro hx
+    simp_all [D.supportWithinDomain hx]
+  · intro hx
+    simp_all
+
+/-- If `U` is closed, the the support of a divisor on `U` is finite. -/
+theorem finiteSupport (D : Divisor U) (hU : IsCompact U) :
+    Set.Finite D.toFun.support :=
+  (hU.of_isClosed_subset (D.closedSupport hU.isClosed)
+    D.supportWithinDomain).finite D.discreteSupport
+
+/-!
+## Lattice ordered group structure
+
+This section equips divisors on `U` with the standard structure of a lattice
+ordered group, where addition, comparison, min and max of divisors are defined
+pointwise.
+-/
+
+/-- Divisors have a zero -/
+instance : Zero (Divisor U) where
+  zero := ⟨fun _ ↦ 0, by simp, Eq.eventuallyEq rfl⟩
+
+/-- Helper lemma for the `simp` tactic: the function of the zero-divisor is the
+zero function. -/
+@[simp]
+theorem zero_fun : (0 : Divisor U).toFun = 0 := rfl
+
+/-- Divisors can be added -/
+instance : Add (Divisor U) where
+  add := by
+    intro D₁ D₂
```

Comment:
```suggestion
  add D₁ D₂ := by
```
Untested, but I think this works.
If it does, then you now have a `by exact` which you can remove and stay in term mode.

Idem dito for `Neg` and `SMul`.

Reply:
Implemented.